src/swarm/neo/util/Batch.d

```diff
@@ -0,0 +1,600 @@
+
+    public bool add ( Record record, void delegate ( ) batch_finished )
+    {
+        if ( !this.fits(record) )
+            return false;
+
+        foreach ( field; record )
+            this.addField(field);
+
+        if ( this.buffer.length >= this.max_size )
+        {
+            batch_finished();
+            this.clear();
+        }
+
+        return true;
+    }
+
+    /***************************************************************************
+
+        Returns:
+            a slice of the data in the batch
+
+    ***************************************************************************/
+
+    public Const!(void)[] get ( )
+    {
+        return *this.buffer;
+    }
+
+    /***************************************************************************
+
+        Compresses the data in the batch using the provided LZO instance and
+        destination buffer.
+
+        Params:
+            lxo = LZO instance to use for compression
+            compress_buf = buffer to receive compressed batch
+
+        Returns:
+            a slice of the compressed data in compress_buf
+
+    ***************************************************************************/
+
+    public Const!(void)[] getCompressed ( Lzo lzo, ref void[] compress_buf )
+    {
+        // Set destination to max possible length.
+        compress_buf.length =
+            lzo.maxCompressedLength(this.buffer.length) + size_t.sizeof;
+        enableStomping(compress_buf);
+
+        // Write uncompressed length into first size_t.sizeof bytes of dest.
+        *(cast(size_t*)(compress_buf.ptr)) = this.buffer.length;
+
+        // Compress into destination.
+        auto dst = compress_buf[size_t.sizeof .. $];
+        auto compressed_len = lzo.compress(*this.buffer, dst);
+
+        // Minimize dest length and return.
+        compress_buf.length = compressed_len + size_t.sizeof;
+        enableStomping(compress_buf);
+        return compress_buf;
+    }
+
+    /***************************************************************************
+
+        Removes all records from the batch.
+
+    ***************************************************************************/
+
+    public void clear ( )
+    {
+        this.buffer.length = 0;
+        enableStomping(*this.buffer);
+    }
+
+    /***************************************************************************
+
+        Calculates the size in the batch required by a single record field.
+
+        Params:
+            Field = type of field
+            field = field to calculate batched size of
+
+        Returns:
+            the number of bytes the specified field would consume in the batch
+
+    ***************************************************************************/
+
+    private size_t sizeOfField ( Field ) ( Field field )
+    {
+        static if ( isDynamicArrayType!(Field) )
+        {
+            static assert(rankOfArray!(Field) == 1,
```

Comment:
`static assert(!hasIndirections!(typeof(field[0])))`, otherwise an array of structs containing arrays would be accepted for `Field`.